lib/metrics/context.js

```diff
@@ -6,43 +6,178 @@
+      events = [ events ]
+    }
+
+    if (! Array.isArray(events)) {
+      return P.resolve()
+    }
+
+    var memcached = getMemcached()
+
+    return P.all(events.map(function (event) {
+      return memcached.setAsync(getKey(token, event), metadata, config.memcached.lifetime)
+        .catch(function (err) {
+          log.error({ op: 'metricsContext.save', err: err })
+        })
+    }))
+  }
+
+  function getMemcached () {
+    try {
+      if (config.memcached.address === 'none') {
+        return NULL_MEMCACHED
       }
+
+      if (! _memcached) {
+        _memcached = new Memcached(config.memcached.address, {
+          timeout: 500,
+          retries: 1,
+          retry: 1000,
+          reconnect: 1000,
+          idle: config.memcached.idle,
+          namespace: 'fxa-metrics~'
+        })
+      }
+
+      return _memcached
+    } catch (err) {
+      log.error({ op: 'metricsContext.getMemcached', err: err })
+
+      return NULL_MEMCACHED
     }
+  }
+
+  /**
+   * Copies metrics context metadata to data, using either metadata
+   * passed in from a request or previously-saved metadata for a
+   * token.
+   *
+   * @param data    target object
+   * @param request request object
+   * @param event   event name
+   */
+  function copy (data, request, event) {
+    var metadata = request.payload && request.payload.metricsContext
+    var token = request.auth && request.auth.credentials
+    var doNotTrack = request.headers && request.headers.dnt === '1'
+
+    return P.resolve()
+      .then(function () {
+        if (metadata) {
+          return metadata
+        }
+
+        if (token && event) {
+          return restore(token, event)
+        }
+      })
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.copy', err: err })
+      })
+      .then(function (metadata) {
+        if (metadata) {
+          data.time = Date.now()
+          data.flow_id = metadata.flowId
+          data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
+          data.context = metadata.context
+          data.entrypoint = metadata.entrypoint
+          data.migration = metadata.migration
+          data.service = metadata.service
+
+          if (! doNotTrack) {
+            data.utm_campaign = metadata.utmCampaign
+            data.utm_content = metadata.utmContent
+            data.utm_medium = metadata.utmMedium
+            data.utm_source = metadata.utmSource
+            data.utm_term = metadata.utmTerm
+          }
+        }
 
-    return data
+        return data
+      })
   }
+
+  function restore (token, event) {
+    var memcached = getMemcached()
+    return memcached.getAsync(getKey(token, event))
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.restore', err: err })
+      })
+  }
+
+  /**
+   * Removes metrics context metadata for a token:event pair.
+   *
+   * @param token
+   * @param event
+   */
+  function remove (token, event) {
```

Comment:
I'm not persuaded about this one yet.

By my count there are 5 calls to `activityEvent` which aren't followed by a call to `metricsContext.remove` and only 3 calls which are. Further, the responsibility for deciding when to call `metricsContext.save` is already inside the route handlers, so it feels proper (to me) that the responsibility for making the matching calls to `metricsContext.remove` should live in the same place.

And if we did move calls to `metricsContext.remove` into `metricsContext.restore`, we'd need to do one of the following:

1. Maintain a list of events that are known to require removal from memcached, or;
2. Pass in a flag to `activityEvent`, indicating whether the event is one that requires removal, or;
3. Do neither and accept that we're making spurious requests to memcached to remove items that we know don't exist.

None of those appeal to me massively. `1` feels like the wrong level of abstraction for that knowledge to live at. `2` feels like unnecessarily passing extra data around when the point of origin could just do it itself. `3` feels mostly harmless but I also don't see what it gains us.

But, that all said, I'm happy to make the change if you really want it. Or maybe I missed something or misunderstood your suggestion, in which case set me straight! 😄